pkg/common-controller/snapshot_controller.go

```diff
@@ -252,6 +269,12 @@ func (ctrl *csiSnapshotCommonController) processSnapshotWithDeletionTimestamp(sn
 	// content has been found from content cache store
 	content, err := ctrl.getContentFromStore(contentName)
 	if err != nil {
+		ctrl.metricsManager.RecordMetrics(metrics.Operation{
+			Name:         deleteSnapshotOperationName,
+			Driver:       driverName,
+			ResourceID:   snapshot.UID,
+			SnapshotType: string(snapshotProvisionType),
+		}, NewSnapshotOperationStatus(snapshotCodeControllerError))
```

Comment:
Recording a metric completion with an error status is mixing the semantic of what this metric is supposed to measure, which is one operation end to end. Instead, what this logic is doing is recording a metric per sync loop, so for a single volume snapshot request, it's could potentially record multiple operations in the histogram for it.

If this is what we want, then we should change the description of the metric to reflect that and look into end to end metrics separately.

For end to end metrics, I'm wondering if actually all we need to do is compare with the object creation/deletion timestamps.

Reply:
The goal of this metric is the following:
>Snapshot operation end to end duration in number of seconds. For example, the duration from a user creates a VolumeSnapshot to request taking a snapshot of a particular PVC to the VolumeSnapshot has a Status.Ready == true.

I like the idea of comparing `VolumeSnapshot.CreationTimestamp` with `VolumeSnapshot.Status.CreationTime` for the e2e operation length. 

We could have two metrics, one counter w/ status, and one e2e with the operation time. This is what I originally implemented. I'll let @xing-yang @yuxiangqian chime in on if just having the histogram is enough. I saved the counter metric code and easily can re-add it if needed.

Reply:
I like to think of it as two different things to measure, and each should have its own set of metrics.

1) Total end to end time of an request. This covers the time from an end user perspective. A controller could retry multiple times, but from the user this is only one request. Intermediate errors are not a concern here.
2) Time of a controller reconciliation loop. This covers just one loop and is more for internal tracking of how efficient this controller is, and can offer some data on intermittent/intermediate errors.

The metric for each scenario could be a histogram that covers status and latency together. There is latency and status of a single control loop, and there is latency and status of a total end to end request (although the status is just success/cancel)

Reply:
That sounds good to me. I'll change the PR soon to reflect the two different histograms.